Section3

GaxysGalactic opened the pull request:
Section 3 Exercises. Please ignore the TestProject folder, it is a testing No Man's Land.

GaxysGalactic committed:
Section3

Section 3 Exercises. Please ignore the TestProject folder, it is a testing No Man's Land.

JDAlzate commented on Section3/Main.cpp:
```diff
@@ -0,0 +1,32 @@
+#include <iostream>
```
Remember to name your projects appropriately. I can tell this is exercise 1 by reading the code, but there's nothing in the name that gives that information.

mcdrakteravision commented:
@GaxysGalactic remember to use .gitignore so that vcxproj files, among others, aren't uploaded to the repository.

JDAlzate commented on Section3/Main.cpp:
```diff
@@ -0,0 +1,32 @@
+	int numberOfAttempts = 0;
+
+	std::cout << "Please enter a number between 1-10: ";
+	std::cin >> std::setw(1) >> choice;
```
Doesn't seem like there's a need to add the `std::setw(1)` part, is there? It's always good to have a reason for everything you do in your code. It's part of understanding what you're doing, rather than just doing what seems to work.

JDAlzate commented on Section3/Main.cpp:
```diff
@@ -0,0 +1,32 @@
+		std::cin.ignore(INT_MAX, '\n');
+
+		std::cout << "Please enter a number between 1-10: ";
+		std::cin >> std::setw(1) >> choice;
```
Same here.

JDAlzate commented on Exercise2/Main.cpp:
```diff
@@ -0,0 +1,100 @@
+
+char getChoice()
+{
+	char choice = -1;
```
Try not to use numeric values when dealing with characters. Try to stick to explicit values of the type you're using, unless you absolutely need to do otherwise.

Though it is generally a good idea to initialize variables when declared, variables that will be assigned with the value of input could be safely left uninitialized. So writing
```
char choice;
std::cin >> choice;
```
would be perfectly safe. But I would recommend having it as close as possible to the initialization code. So instead of 
```
char choice;
int currentAttempt = 1;
std::cin >> choice;
```
you could instead define `currentAttempt` before declaring the variable `choice` or after calling the `std::cin >> choice` line. That way, you're not leaving any room for modification of your code between the declaration and definition of `choice`.

JDAlzate commented on Exercise2/Main.cpp:
```diff
@@ -0,0 +1,100 @@
+{
+	char choice = -1;
+	int currentAttempt = 1;
+	std::cin >> std::setw(1) >> choice;
```
Please make sure to remove all these std::setw from your code, unless there is a reason to have them :)

JDAlzate commented on Exercise3/Main.cpp:
```diff
@@ -0,0 +1,20 @@
+		}
+		else
+		{
+			sum += i;
```
The exercise asked you to not use an else statement, please fix that.

JDAlzate commented on Exercise4/Main.cpp:
```diff
@@ -0,0 +1,11 @@
+	std::cout << "Multiples of 10: " << std::endl;
+	for (int i = 1; i <= 10; i++)
+	{
+		int mult = 10 * i;
```
The code is good and the output is correct! What I wanted for you to do was to change the `i++` for `i+10` and the `i <= 10` to `i <= 100`, but I never specified that you could not perform multiplication in the exercise, so it's all good! Just saying so you know what the intention of the exercise was.

jeroo97 commented on Exercise3/Main.cpp:
```diff
@@ -0,0 +1,20 @@
+		{
+			std::cout << i << " ";
+		}
+		else
```
Remember, in this exercise, you shouldn't use else statements. Maybe you could try a break statement somewhere?

JDAlzate commented on Exercise5/Main.cpp:
```diff
@@ -0,0 +1,36 @@
+	std::cout << "Enter a number if you want: ";
+	std::cin >> std::setw(1) >> number;
+
+	int i;
```
If you're going to have your variable outside of your loop, keep in mind that it's perfectly fine to have the declaration parto of the for loop empty. So you can write:
```
int i = 0;
for(; std::cin.good() != false && i < 10; i++)
{
    // blah blah
}
```

That way you don't have an uninitialized variable more than you need to.

JDAlzate replied:
Also, keep in mind that expressions that evaluate to a boolean can be directly checked as true or false typing `if  (std::cin.good())` or `if (!std::cin.good())` adding a `== true` or `== false` is a bit redundant, and more so if it's negated like you have it here. An `if (std::cin.good() != false)` may be harder to read than `if (std::cin.good() == true)` which is redundant compared to `if (std::cin.good())`

JDAlzate commented on Exercise6/Main.cpp:
```diff
@@ -0,0 +1,43 @@
+{
+	int myBetterArray[10];
+	int number = -1;
+	int i;
```
There is no gain in declaring variables before anything else in your code. The way to make your code the most readable is by declaring your variables just before you need them so that they make sense in the context you find them in. Having this `i` variable at the top of the main function, for example, does nothing  to me as the reader but wonder what i means, while having it inside or just before the for loop makes it clear that it is an index.

Another reason to do this is for scope reduction. You are defining the `number`` variable outside of the scope of the for loop, but the variable is only ever used inside that scope, so there is no reason to have access to it from anywhere else.

JDAlzate commented on Exercise6/Main.cpp:
```diff
@@ -0,0 +1,43 @@
+	{
+		std::cout << "Enter a number if you want or press ENTER to finish: ";
+
+		if (std::cin.peek() == 10) break;
```
Why are you checking against 10? Is 10 the numeric value for a new line? If that's the case, wouldn't it be easier to read if you compared directly against a character rather than it's numeric representation?

JDAlzate commented on Exercise6/Main.cpp:
```diff
@@ -0,0 +1,43 @@
+		if (i == 9)
+		{
+			std::cout << "I did not account for this. Array is now full." << std::endl;
+			continue;
```
Is there a need for this `continue` here? It all comes down to what I said before, always try to have a reason for everything you're doing.

jeroo97 commented on Exercise5/Main.cpp:
```diff
@@ -0,0 +1,36 @@
+	int myFirstArray[10];
+	int number = -1;
+
+	std::cout << "Enter a number if you want: ";
```
You could place these two lines at the beginning of the for loop. That way you don't need to repeat them in lines 32 & 33

JDAlzate commented on Exercise7/Main.cpp:
```diff
@@ -0,0 +1,73 @@
+
+bool hasRepeats(int list[])
+{
+	for (int i = 1; i < 5; i++)
```
You did well on the first function, passing in the size of the array. However, you didn't do that here. 

Hardcoding values is error-prone. A function is written to be useful for different inputs, so you're allowing the function to be called with an array but nowhere in the function is it specified that the list should contain exactly 5 items.

JDAlzate commented on Exercise9/Main.cpp:
```diff
@@ -0,0 +1,36 @@
+
+	int size;
+	for (size = 0; myString[size] != '\0'; size++);
+	size;
```
What's the need for this line?

JDAlzate commented on Exercise9/Main.cpp:
```diff
@@ -0,0 +1,36 @@
+	for (int i = 0; i < size; i++)
+	{
+		myReversedString[size - i - 1] = myString[i];
+		if (myString[i + 1] == '\0')
```
You're checking that you're in the last character of your string to assign the \0 to your `myReversedString[size]`. This will always happen at the end of the loop, so there is no need to actually do it inside the loop. Why not just move the line `myReversedString[size] = \0` to be called after the loop? That way you can eliminate the condition.

Always be mindful of what needs and doesn't need to be inside a for loop.

JDAlzate reviewed (changes requested):
Review is up!

GaxysGalactic committed:
Review Changes

- [FIX] Bug fixes on input validation for lottery
- [FIX] Changes based on code review
- [NEW] Do/While loops to avoid rewriting code in some input validation exercises
- [NEW] GitIgnore file with proper ignoring, though a bit too late for that.

JDAlzate reviewed (approved):
Approved! :)
